framework/Util/Cron/TTimeScheduler.php

```diff
@@ -0,0 +1,856 @@
+					$data['end'] = 6; //No end with period
+				}
+				$this->_attr[self::DAY_OF_WEEK][] = $data;
+			} else {
+				throw new TInvalidDataValueException('timescheduler_invalid_string', $match);
+			}
+		}
+			
+		$this->_attr[self::YEAR] = [];
+		$matches[7] = $matches[7] ?? '*';
+		foreach (explode(',', $matches[7]) as $match) {
+			if (preg_match('/^(\*|' . $year . ')(?:\-(' . $year . '))?(?:\/([1-9]?[0-9]))?$/i', $match, $m2)) {
+				if ($m2[1] === '*') {
+					$data = ['year' => self::YEAR_MIN];
+					$data['end'] = self::YEAR_MAX;
+				} else {
+					$data = ['year' => $m2[1]];
+					$data['end'] = $m2[2] ?? $m2[1];
+				}
+				$data['period'] = max($m2[3] ?? 1, 1);
+				if (!($m2[2] ?? 0) && ($m2[3] ?? 0)) {
+					$data['end'] = self::YEAR_MAX; //No end with period
+				}
+				$this->_attr[self::YEAR][] = $data;
+			} else {
+				throw new TInvalidDataValueException('timescheduler_invalid_string', $match);
+			}
+		}
+	}
+	
+	protected function translateMonthOfYear($v)
+	{
+		if (is_numeric($v)) {
+			return $v;
+		}
+		foreach (self::$_keywords[self::MONTH_OF_YEAR] as $moy => $regex) {
+			if (preg_match('/^(' . $regex . ')$/i', $v)) {
+				return $moy;
+			}
+		}
+		return $v;
+	}
+	
+	protected function translateDayOfWeek($v)
+	{
+		if (is_numeric($v)) {
+			return $v;
+		}
+		foreach (self::$_keywords[self::DAY_OF_WEEK] as $dow => $regex) {
+			if (preg_match('/^(' . $regex . ')$/i', $v)) {
+				return $dow;
+			}
+		}
+		return $v;
+	}
+	
+	
+	/**
+	 * Given which minutes are valid for the parsed time stamp, this generates
+	 * an array of valid minutes.  Each minute is available as the key to the array
+	 * and the value contains true or false whether the minute meets the cron
+	 * time stamp
+	 */
+	protected function getMinutesArray()
+	{
+		$ma = array_pad([], 60, null);
+		foreach ($this->_attr[self::MINUTE] as $m) {
+			if (is_numeric($m['min'])) {
+				for ($i = $m['min']; $i <= $m['end'] && $i < 60; $i += $m['period']) {
+					$ma[$i] = 1;
+				}
+			}
+		}
+		return $ma;
+	}
+	
+	
+	/**
+	 * Given which hours are valid for the parsed time stamp, this generates
+	 * an array of valid hours.  Each hour is available as the key to the array
+	 * and the value contains true or false whether the hour meets the cron
+	 * time stamp
+	 */
+	protected function geHoursArray()
+	{
+		$ha = array_pad([], 24, null);
+		foreach ($this->_attr[self::HOUR] as $h) {
+			if (is_numeric($h['hour'])) {
+				for ($i = $h['hour']; $i <= $h['end'] && $i < 24; $i += $h['period']) {
+					$ha[$i] = 1;
+				}
+			}
+		}
+		return $ha;
+	}
+	
+	
+	/**
+	 * Returns the number of days in a given month and year, taking into account leap years.
+	 *
+	 * @param $month: numeric month (integers 1-12)
+	 * @param $year: numeric year (any integer)
+	 */
+	public function days_in_month($month, $year)
+	{
+		return $month == 2 ? ($year % 4 ? 28 : ($year % 100 ? 29 : ($year % 400 ? 28 : 29))) : (($month - 1) % 7 % 2 ? 30 : 31);
```

Comment:
This could be done with
```php
cal_days_in_month(CAL_GREGORIAN, $month, $year)
```
But i guess you avoided that method since it requires the calendar module.
As an alternative you can use this:
```php
date('t', mktime(0, 0, 0, $month, 1, $year)); 
```
I'm not sure if this really makes any difference in terms of performance, so feel free to use the approach that you prefer.

Reply:
100.000 times and here are the results:
2.5287721157074  for cal_days_in_month
1.4125881195068  for date takes
0.017313957214355  for current solution